Un grupo de investigación en Biología Marina está realizando mediciones en un molusco en el cual están interesados, estos datos los están almacenando en una base de datos desde hace ya un tiempo atrás (abalone.csv).

La información que guardan es la relacionada a:  Sex, Length, Diameter, Height, Whole_weight, Shucked_weight, Viscera_weight y Rings.

Pero últimamente han recolectado datos de los cuales no han podido determinar el Sex del molusco.

Decidieron hacer un proceso de clasificación con los datos que si conocen para calcular el valor de Sex  faltante.

En el proceso de clasificación es probable que tengan que validar datos no capturados e imputarlos con los valores promedio o el valor más frecuente de cada columna y mostrar gráficamente como están los datos antes y después de la imputación, también decidieron hacer un histograma por cada columna.

También puede contener caracteres no validos en columnas numéricas.

Una vez teniendo listos los datos para la clasificación, decidieron probrar con varios métodos.

Encontrado el mejor modelo de clasificación de alguna de estas técnicas, tomarán ese mejor modelo para hacer la clasificación del valor Sex  en el archivo (abalone_missing.csv) que puede tener valores faltantes las columnas o en su caso puede contener caracteres no validos a en columnas numéricas. (Los valores faltantes de Sex, no deben de ser removidos).

# Regresion - abalone.csv

In [81]:
import pandas as pd
import numpy as np
from sklearn.impute import SimpleImputer
from sklearn.linear_model import Lasso, Ridge, LinearRegression, ElasticNet, LogisticRegression
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error,mean_squared_error


archivo="/content/drive/MyDrive/datasets/abalone.csv"
df=pd.read_csv(archivo,sep=",",header=0) 

le=LabelEncoder()
df["sex_le"]=le.fit_transform(df["Sex"])

# ANTES DE LA IMPUTACION Y LIMPIEZA
#ax=msno.bar(df)
#ax.get_figure().show()

#imputacion
si=SimpleImputer(missing_values=np.nan, strategy="mean")
df.sex_le=si.fit_transform(df.sex_le.values.reshape(-1,1)) 


#LIMPIEZA 
df=df.replace({"Diameter":{".":0}}) 
df["Diameter"] = df["Diameter"].astype("float64")

df=df.replace({"Shucked_weight":{"a":0, "s":0}}) 
df["Shucked_weight"] = df["Shucked_weight"].astype("float64")

df=df.replace({"Viscera_weight":{"a":0, "s":0, "e":0, "r":0,"w":0,"f":0,"g":0,"h":0,"u":0,"df":0}})
df["Viscera_weight"] = df["Viscera_weight"].astype("float64")

df=df.replace({"Shell_weight":{"a":0, "s":0}}) 
df["Shell_weight"] = df["Shell_weight"].astype("float64")

df=df.fillna(0)

#DESPUES DE LA IMPUTACION Y LIMPIO
#ax=msno.bar(df)
#ax.get_figure().show()


#ESCOGIENDO EL MEJOR MODELO ENTRENADO
features=["Length","Diameter", "Height", "Whole_weight", "Shucked_weight", "Viscera_weight","Rings"]

train,test, train_label, test_label = train_test_split(df[features],df["sex_le"],random_state=0, train_size=0.30)

lr=LogisticRegression()
modelo= lr.fit(train,train_label)
prediccion= modelo.predict(test)

prediccion_list=prediccion.tolist() 
test_label_list=test_label.tolist()

score= modelo.score(test,test_label)
mae= mean_absolute_error(prediccion,test_label) 
mse= mean_squared_error(prediccion,test_label) 

#score= modelo.score(test,test_label) 
#acs= accuracy_score(prediccion,test_label)  
#cm= confusion_matrix(prediccion,test_label) 
#cr= classification_report(prediccion,test_label) 

#print("score= ", score, "\naccuracy_score=", acs,"\nconfusion_matrix=", cm,"\nclassification_report=",cr)
print("score= ", score, "\nmedia error absoluto=", mae,"\nmedia error cuadrado=", mse)


score=  0.5557455540355677 
media error absoluto= 0.6939124487004104 
media error cuadrado= 1.1932284541723666


/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Resultados obtenidos:

## Mejor modelo de regresion: Logistic Regression

*   LinearRegression: 

> score=  0.0068677723838376226 

*   Ridge:

> score=  0.006645628898964473


*   Lasso:

> score=  -8.627955343820659e-05


*   ElasticNet:

> score=  -8.627955343820659e-05


*   LogisticRegression:

> score=  0.5557455540355677






# Clasificacion - abalone.csv

In [85]:
import pandas as pd
import numpy as np
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, f1_score, classification_report


archivo="/content/drive/MyDrive/datasets/abalone.csv"
df=pd.read_csv(archivo,sep=",",header=0) 

le=LabelEncoder()
df["sex_le"]=le.fit_transform(df["Sex"])

# ANTES DE LA IMPUTACION Y LIMPIEZA
#ax=msno.bar(df)
#ax.get_figure().show()

#imputacion
si=SimpleImputer(missing_values=np.nan, strategy="mean")
df.sex_le=si.fit_transform(df.sex_le.values.reshape(-1,1)) 


#LIMPIEZA 
df=df.replace({"Diameter":{".":0}}) 
df["Diameter"] = df["Diameter"].astype("float64")

df=df.replace({"Shucked_weight":{"a":0, "s":0}}) 
df["Shucked_weight"] = df["Shucked_weight"].astype("float64")

df=df.replace({"Viscera_weight":{"a":0, "s":0, "e":0, "r":0,"w":0,"f":0,"g":0,"h":0,"u":0,"df":0}})
df["Viscera_weight"] = df["Viscera_weight"].astype("float64")

df=df.replace({"Shell_weight":{"a":0, "s":0}}) 
df["Shell_weight"] = df["Shell_weight"].astype("float64")

df=df.fillna(0)

#DESPUES DE LA IMPUTACION Y LIMPIO
#ax=msno.bar(df)
#ax.get_figure().show()


#ESCOGIENDO EL MEJOR MODELO ENTRENADO
features=["Length","Diameter", "Height", "Whole_weight", "Shucked_weight", "Viscera_weight","Rings"]

train,test, train_label, test_label = train_test_split(df[features],df["sex_le"],random_state=0, train_size=0.30)

lr=GradientBoostingClassifier()
modelo= lr.fit(train,train_label)
prediccion= modelo.predict(test)

prediccion_list=prediccion.tolist() 
test_label_list=test_label.tolist()


score= modelo.score(test,test_label) 
acs= accuracy_score(prediccion,test_label)  
cm= confusion_matrix(prediccion,test_label) 
cr= classification_report(prediccion,test_label) 

print("score= ", score, "\naccuracy_score=", acs,"\nconfusion_matrix=", cm,"\nclassification_report=",cr)


score=  0.5434336525307798 
accuracy_score= 0.5434336525307798 
confusion_matrix= [[312  70 303]
 [117 700 188]
 [491 166 577]] 
classification_report=               precision    recall  f1-score   support

         0.0       0.34      0.46      0.39       685
         1.0       0.75      0.70      0.72      1005
         2.0       0.54      0.47      0.50      1234

    accuracy                           0.54      2924
   macro avg       0.54      0.54      0.54      2924
weighted avg       0.56      0.54      0.55      2924




## Mejor modelo de clasificacion: GradientBoostingClassifier

*   DecisionTreeClassifier:
> score=   0.4904240766073871 


*   RandomForestClassifier
> score=  0.5294117647058824 


*   GradientBoostingClassifier
> score=  0.5434336525307798





# Clasificacion - abalone_missing.csv

In [126]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, f1_score, classification_report

archivo="/content/drive/MyDrive/datasets/abalone_missing.csv"
df=pd.read_csv(archivo,sep=",",header=0) 



df=df.fillna(0) #Rellenar datos nulos 

#LIMPIEZA 
df=df.replace({"Diameter":{"g":0}}) 
df["Diameter"] = df["Diameter"].astype("float64")

df=df.replace({"Length":{"u":0}}) 
df["Length"] = df["Length"].astype("float64")

df=df.replace({"Height":{"s":0, "hre":0, "r":0, "u":0,"fg":0}}) 
df["Height"] = df["Height"].astype("float64")

df=df.replace({"Whole_weight":{"g":0, "s":0, "jk":0, "u":0,"fg":0}}) 
df["Whole_weight"] = df["Whole_weight"].astype("float64")

df=df.replace({"Viscera_weight":{"asd":0, "d":0, "s":0, "f":0,"i":0}}) 
df["Viscera_weight"] = df["Viscera_weight"].astype("float64")



#ESCOGIENDO EL MEJOR MODELO ENTRENADO
features=["Length","Diameter", "Height", "Whole_weight", "Shucked_weight", "Viscera_weight","Rings"]

train,test, train_label, test_label = train_test_split(df[features],df["Sex"],random_state=0, train_size=0.30)

lr=GradientBoostingClassifier()
modelo= lr.fit(train,train_label)
prediccion= modelo.predict(test)

prediccion_list=prediccion.tolist() 
test_label_list=test_label.tolist()


score= modelo.score(test,test_label) 
acs= accuracy_score(prediccion,test_label)  
cm= confusion_matrix(prediccion,test_label) 
cr= classification_report(prediccion,test_label) 

print("score= ", score, "\naccuracy_score=", acs,"\nconfusion_matrix=", cm,"\nclassification_report=",cr)

ValueError: ignored